### PR TITLE
fix(capture): strip invalid utf16 surrogate pairs from input

### PR DESCRIPTION
**Pull request by oliverb123**
## Problem

The browser (or rrweb?) emits invalid UTF16 surrogate pairs in json strings for session replay sometimes, and we reject that outright in capture. This PR is the second attempt at, instead, replacing those invalid pair escape sequences with valid ones, prior to deserialising the json.

I'm marginally higher-confidence with this attempt than the previous one, although I'm still a little 🤷 - the semantics here are json/js-specific (I spent far too long reading about the distinction between DOMString, JsonString and `scalar string values` trying to figure out what the "right" behaviour here is), and a bit gnarly.

The major reason I'm higher confidence is that the previous attempt was working on a per-byte basis, which in hindsight, was an obvious mistake, and led to utf8 parsing errors at the end of the algorithm pass. This implementation relies on rust's `char` type, which represents a valid, variable-sized utf8 codepoint - and strictly consumes and emits those, which means it'll never emit a character sequence that isn't a valid utf8 `String` (note we're never calling an `String::parse` method).

### CODE REVIEW

File: rust/capture/src/passes/invalid_surrogates.rs [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+                self.last_seen = LastSeen::Char;
+                return Some('\\');
+            };
+            return None;
+        };
+
+        match (self.last_seen, c) {
+            (LastSeen::Escape, 'u') => {
+                self.escape_seq_buf.clear();
+                if collect_n_chars(&mut self.escape_seq_buf, &mut self.input, 4) < 4 {
```
Comment (bretthoerner) [inline rust/capture/src/passes/invalid_surrogates.rs:109]: Would be nice to have a like `HEX_ESCAPE_LENGTH = 4` const if that's what the 4 is represending?

File: rust/capture/src/passes/invalid_surrogates.rs [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.step()
+    }
+}
+
+impl<'a> InvalidSurrogatesPass<'a> {
+    // This is a simple heuristic to determine if we need to run this pass
+    pub fn needed(input: &str) -> bool {
+        input.contains("\\u")
```
Comment (bretthoerner) [inline rust/capture/src/passes/invalid_surrogates.rs:33]: I know it's simple, but 2 tiny test cases to verify true/false where expected would be nice... this is the kind of thing someone will change on a whim one day and break everything without knowing.

**Review by bretthoerner — APPROVED** [review]
At a pass, it overall makes sense to me. What would make me feel a lot better is more test cases that try to do both normal and weird things.

Or... I am just lacking context, I don't know how/why the last attempt failed. Does the 1 test case here cover it at least?

**Comment by oliverb123** [comment]
@bretthoerner the last attempt failed mostly due to `String` parsing failures because I was operating on them as byte-slices and mangling the UTF8 - which is a class of errors that simply can't happen here. 

You're right about tests though - I had taken the view "this is in the deserialisation path, all the existing tests hit it", but because I do the "is needed" check, they mostly actually don't. I'm going to remove that (in the unneeded cases, this is still a fast pass over the string, even if it's a bit inefficient, we're doing silly stuff anyway here), and then the "happy path" will be covered heavily by existing tests.

File: rust/capture/src/passes/invalid_surrogates.rs (hunk 2 above)
Comment (oliverb123) [reply rust/capture/src/passes/invalid_surrogates.rs]: Removed instead, premature opt

File: rust/capture/src/passes/invalid_surrogates.rs (hunk 1 above)
Comment (oliverb123) [reply rust/capture/src/passes/invalid_surrogates.rs]: Done

**Comment by oliverb123** [comment]
I kind of hate that this code exists, but I also don't know of a better way to handle this - I think it's good we're no longer passing invalid utf8/16 downstream (even if plugin-server can handle it, it's better to be stricter here for the sake of future systems - who knows what will be reading session data in the future, and how it would handle this). There is also a degree to which this is captures _whole job_ - data validation. I've thought about putting that "remove nulls" code we have littered around the place here too, as another pass - maybe a TODO for the future.

If I could I would not merge this, but I think it's a necessary evil.

**Pull request merged**